---
title: "Managing multiple data-frames"
subtitle: "Presented to EdinbR R-Users Group, 2018-07-18"
author: "Russ Hyde, University of Glasgow"
date: "`r Sys.Date()`"
output:
  ioslides_presentation:
    df_print: paged
---

<!--
  Styling to allow footnotes to appear on the same slide as their definition
  - Borrowed from https://stackoverflow.com/questions/42690955
-->

<style>
div.footnotes {
  position: absolute;
  bottom: 0;
  margin-bottom: 10px;
  width: 80%;
  font-size: 0.6em;
}
</style>

<script
  src="https://ajax.googleapis.com/ajax/libs/jquery/3.1.1/jquery.min.js"
  ></script>

<script>
  $(document).ready(function() {
    $('slide:not(.backdrop):not(.title-slide)').append('<div class=\"footnotes\">');

    $('footnote').each(function(index) {
      var text  = $(this).html();
      var fnNum = (index+1).toString().sup();
      $(this).html(text + fnNum);

      var footnote   = fnNum + ': ' + $(this).attr('content') + '<br/>';
      var oldContent = $(this).parents('slide').children('div.footnotes').html();
      var newContent = oldContent + footnote;
      $(this).parents('slide').children('div.footnotes').html(newContent);
    });
  });
</script>

<!-- Packages -->

```{r, echo = FALSE}
suppressPackageStartupMessages({
  library(Biobase)
  library(DiagrammeR)
  library(dplyr)
  library(ggplot2)
  library(igraph)
  library(igraphdata)
  library(pryr)
  library(purrr)
  library(reshape2)
  library(stringr)
  library(tibble)
  library(tidyr)
  library(tidygraph)
})
```

<!-- Presentation code -->

## Background and Links:

- @POG_LRC / @GUcancersci / @bloodwise_uk
    - I'm a postdoc bioinformatician at The Paul O'Gorman (POG) Leukaemia
    Research Centre (University of Glasgow)
    - ... working for Prof. Mhairi Copland (POG) and Dr. David Vetrie
    (Wolfson-Wohl Cancer Research Centre)
    - ... on a Bloodwise-funded grant
    - ... into chronic-myeloid leukaemia

- @haematobot
    - Personal mumblings about code / analysis / bioinformatics and seemingly
    very little else ...

- https://biolearnr.blogspot.com/
    - Even more mumblings

## Preamble

See `https://github.com/russHyde/polyply`

```{r}
# Dependencies:
# - purrr, methods, rlang, tidygraph, dplyr
if (!"polyply" %in% installed.packages()) {
  require("devtools")
  devtools::install_github(
    repo = "russHyde/polyply", dependencies = FALSE
  )
}

suppressPackageStartupMessages(
  library(polyply)
)
```

# Data-Modelling

## Tidy Data and the Normal-Forms {.build}

In <footnote content="http://vita.had.co.nz/papers/tidy-data.html">tidy data</footnote>:

- TD1 - Each variable forms a column.
- TD2 - Each observation forms a row.
- TD3 - Each type of observational unit forms a table.
- [TD4 - A key permitting table-joins is present]

See also, Boyce-Codd
<footnote content="https://en.wikipedia.org/wiki/First_normal_form">Normal-Forms</footnote> and relational-database-design.

- ?? TD5 - A tidy way of encapsulating your nicely decomposed tables
- ?? TD6 - An explicit workflow for combining your tables back together

## Common _Untidy_ Data Structures

Tidy-data / normal-forms in R

- $\downarrow$ duplication

- play nicely with some important things (`ggplot2` etc)

But untidy data-structures are useful if they:

- $\uparrow$ access efficiency

- $\downarrow$ code complexity

- play nicely with other important things

## `Biobase::ExpressionSet`

```{r, message=FALSE, eval=FALSE}
Biobase::ExpressionSet()
```

```{r, echo = FALSE}
grViz("
digraph eset_boxes {

  # graph, node, and edge definitions
  graph [compound = true, nodesep = .5, ranksep = .25,
         color = crimson]

  node [fontname = Helvetica, fontcolor = darkslategray,
        shape = rectangle, fixedsize = true, width = 3,
        color = darkslategray]

  edge [color = grey, arrowhead = none, arrowtail = none]

  subgraph eset {
    # nodes
    node [fixedsize = true]
    A [label = 'assayData\n(matrix)\n\nnrow=|genes|\nncol=|samples|', height = 3, width = 2]
    B [label = 'featureData\n(data-frame)\n\nnrow=|genes|', height = 3, width = 2.5]
    C [label = 'phenoData\n(data-frame)\n\nnrow=|samples|', height = 2, width = 2.5]

    # edges
    A->B A->C
  }

  ExpressionSet         [width = 2]
  ExpressionSet -> A    [lhead = eset]
  ExpressionSet -> protocolData
  ExpressionSet -> experimentData
  ExpressionSet -> Annotation
}
")
```

Figure made with `DiagrammeR`

## `Biobase::ExpressionSet` (cont.)

Conversion of the `assayData` to meet tidy-data standards:

```{r, echo = FALSE}
m <- as.matrix(data.frame(
  sample1 = c(12.2, 19.1, 0.5),
  sample2 = c(111, 10.5, 3.4),
  sample3 = c(129, 123, 1.1),
  row.names = paste0("gene", 1:3)
))
```

```{r}
m # our assayData
```

Doesn't meet tidy-data standards:

- rows correspond to features, columns to samples
- not all variables are in columns (since row-IDs are meaningful)
- entries are the same 'type' of variable

----

Easy <footnote content="... or as.data.frame / rownames_to_column / gather">fix</footnote>:

```{r}
m2 <- reshape2::melt(
  m,
  varnames = c("feature_id", "sample_id"),
  as.is = TRUE
)

head(m2, 4)
```

## But ...

- Matrix representation was more dense

- Lost all encapsulation

- (After modifying featureData / phenoData to match)
    - Have to join rather than index
    - Have to keep track of multiple data-frames, rather than one
      data-structure

<!-- Remove this next section: -->

## That multi-data-frame _thing_

For a reasonably complex project:

- tidy-data / normal-forms mean more data-frames

Wanted:

- a lightweight approach to working with multiple 'conceptually-related'
data-frames

- that plays nicely with `tidyverse` verbs

- that feeds into `ggplot2`

- that plays nicely with untidy data-structures I use _all the time_

# `tidygraph` already (sort of) does this

## Graph theory

```{r, echo = FALSE}
data(Koenigsberg, package = "igraphdata")
plot(Koenigsberg)
```

## Basics of 'graph theory' speak

A graph is made up of two sets:

- _V_, a set of vertices:
    - aka nodes, actors, ...

- _E_, a set of edges:
    - pairwise relationships between vertices
    - aka interactions, lines, arcs, ...

- Need to store attributes for both nodes and edges

## `tbl_graph` data structure

`tidygraph` is really a wrapper around the package `igraph`

```{r}
data("Koenigsberg", package = "igraphdata")
tg <- tidygraph::as_tbl_graph(Koenigsberg)

# Nodes data shows up first:
tg
```

## `tbl_graph` data structure

```{r}
# If we make the 'edges' active, the edge-data shows up first:
activate(tg, edges)
```

## The `activate` verb

Think of the `tbl_graph` as `list[nodes, edges]`

To modify the contents of a given data-frame, `activate` it:

```{r}
tg %>%
  activate(edges) %>%
  mutate(weight = nchar(name))
```

# `polyply` and multiple, linked data-frames

## `polyply` {.build}

Aim:

- multiple data-frames in one data-structure

    - $\rightarrow$ class `poly_frame`: extends list`
    - `poly_frame`: [list[data-frame], merge_fn]

- mutation / filtering

- merging

## Exported functions

- `as_poly_frame`

    - convert a data-structure into a `poly_frame`

- `activate`

    - choose a data-frame from within the `poly_frame`

- `filter`

    - modify the contents of the active data-frame

- `merge`

    - user defined data-frame combiner (default: reduce(inner_join)(df_list))

- others to be added (mutate / select etc)

# Examples

## ExpressionSet Example

```{r, echo = FALSE}
use_leuk_example <- "leukemiasEset" %in% installed.packages()
```

```{r, eval=use_leuk_example}
data("leukemiasEset", package = "leukemiasEset")
leuk <- leukemiasEset
leuk
```

```{r, eval = !use_leuk_example}
# To see the leukaemia example, install `leukemiasEset` from `bioconductor`
```

## Construct a poly-frame from an ExpressionSet

```{r, eval = use_leuk_example}
leuk_pf <- list(
  exprs = reshape2::melt(
    exprs(leuk),
    as.is = TRUE,
    varnames = c("feature_id", "sample_id")
  ),
  pheno = tibble::rownames_to_column(
    phenoData(leuk)@data,
    var = "sample_id"
  )
) %>%
  as_poly_frame()
```

## What did we just make?

```{r, eval=use_leuk_example}
purrr::map(leuk_pf, head)
```

## Filter and plot:

<!-- TODO: show how ugly this code is within standard tidyverse / base R -->

```{r, eval=use_leuk_example}
my_plot <- leuk_pf %>%
  # At first, data-frame `exprs` is active
  filter(feature_id %in% c("ENSG00000000003", "ENSG00000000005")) %>%
  # Select a different data-frame for filtering:
  # - you can use non-standard-evaluation in `activate`
  activate(pheno) %>%
  # only look at myeloid leukaemias
  filter(LeukemiaType %in% c("AML", "CML")) %>%
  # default merge: fold an inner-join
  merge() %>%
  ggplot()
```

## Filter and plot(cont.)

```{r, fig.height = 3, eval=use_leuk_example}
my_plot +
  geom_boxplot(aes(x = LeukemiaType, y = value)) +
  facet_wrap(~ feature_id) +
  ggtitle("These might not be the most interesting genes in the dataset ...")
```

## Taxonomy and brains

```{r}
data(animals, package = "polyply")
head(animals)
```

## Taxonomies (cont.)

```{r}
data(common_to_species, package = "polyply")
head(common_to_species)
```

## Taxonomies (cont.)

```{r, message=FALSE, warning=FALSE}
data(taxonomy, package = "polyply")
head(taxonomy)
```

## Taxonomies & brains (cont.)

```{r, fig.height = 3}
as_poly_frame(
  list(animals, common_to_species, taxonomy)
) %>%
  merge() %>%
  ggplot(aes(x = body, y = brain, col = order)) +
  geom_point() +
  xlim(0, NA) + ylim(0, NA)
```


# Thanks
